python_bindings/tutorial/lesson_10_generators_1.py

```diff
@@ -0,0 +1,242 @@
+    y = hl.Var("y")
+
+    def generate(self):
+        # Grab the width and height of the input, for future use
+        w = self.input_buffer.width()
+        h = self.input_buffer.height()
+
+        # Define the brightening Func.
+        brighter = hl.Func("brighter")
+        brighter[self.x, self.y] = hl.cast(hl.UInt(8), hl.min(
+            self.input_buffer[self.x, self.y] * self.brighten_factor, 255))
+
+        # Now, possibly do some sort of rotation, depending on the
+        # value specified in self.rotation. (In production code, we'd likely prefer
+        # to create a real Enum class for rotation, but for this simple example,
+        # we'll just use if-elif.)
+        rotated = hl.Func("rotated")
+        if self.rotation == "none":
+            rotated[self.x, self.y] = brighter[self.x, self.y]
+        elif self.rotation == "cw":
+            rotated[self.x, self.y] = brighter[w - self.y - 1, self.x]
+        elif self.rotation == "ccw":
+            rotated[self.x, self.y] = brighter[self.y, h - self.x - 1]
+        else:
+            # Illegal or unsupported values for a GeneratorParam should always be
+            # handled by raising an exception.
+            raise ValueError(
+                "Unsupported value for GeneratorParam rotation: %s" % self.rotation)
+
+        # Finally, assign to the output buffer.
+        self.output_buffer[self.x, self.y] = hl.cast(
+            self.output_buffer.type(), rotated[self.x, self.y])
+
+        # The structure of the pipeline depended on the generator
+        # params, and (in this case) so will the schedule.
+
+        # Let's start by vectorizing the output. We don't know the
+        # type though, so it's hard to pick a good factor. Generators
+        # provide a helper method called "natural_vector_size" which will
+        # pick a reasonable factor for a numeric type, based on the
+        # Target you're compiling to.
+        v = self.natural_vector_size(self.output_buffer.type())
+        self.output_buffer.vectorize(self.x, v)
+
+        # Now we'll possibly parallelize it.
+        if self.parallel:
+            self.output_buffer.parallel(self.y)
+
+        # If there was a rotation, we'll schedule that to occur per
+        # scanline of the output and vectorize it according to its
+        # type.
+        if (self.rotation != "none"):
+            v = self.natural_vector_size(rotated.type())
+            rotated.compute_at(self.output_buffer, self.y).vectorize(self.x, v)
+
+
+def use_second_generator():
+    # We can override values for GeneratorParams by specifying an optional by-name-only argument,
+    # generator_params, to the .call() method. It takes a simple Python dict as its value.
+
+    # We'll read in an image here so that the effects of the code are easier to see.
+
+    im = imageio.imread(os.path.join(os.path.dirname(__file__), "images/gray.png"))
+    # We need to ensure that the image is arranged in 'F' order, so that the innermost axis
```

Comment:
I think this paragraph is potentially a big point of confusion. Normally I use the default C order for numpy arrays and just expect the indexing to be reversed between Halide vs numpy, so numpy_array[y, x] == halide_buffer[x, y], in the same way that we have c_array[y][x] == halide_buffer[x, y] when using C.

IIUC, F order only changes how the indices are interpreted when using numpy's operator[], so this is unrelated to whether or not things vectorize cleanly in Halide, unless we're matching indexing order between the two, which we don't do in C.

Reply:
Specifically, I think this version is confusing because people expect to write numpy_array[row, col], and in lesson 1 we explain that in Halide we index as [col, row]. If they try to index the numpy_array as [col, row] here as they're used to, they'll get the transpose of what they want.

Reply:
> this paragraph is potentially a big point of confusion

I don't disagree, but am not sure what better options we have:
- because of the rotation, I wanted an image that would be obvious from inspection that it 'worked'
- imageio.imread() apparently insists on reading grayscale images (or at least *this* one) gets read with stride=1 on the outmost axis rather than the innermost, so we can't vectorize() in the usual way with this image
- I suppose I could upgrade to the RGB image, it would just make the example more verbose?
- Or maybe just construct a buffer in-memory and draw a simple arrow or gradient on it?

Reply:
imread loads an image such that the first row of the image is dense in memory.

Halide with default memory layouts and indexing conventions also expects the first row of the image to be dense in memory.

These should be compatible. My suspicion is that the numpy array -> Halide buffer conversion is converting the numpy strides/sizes arrays in order whereas it should be flipping the order.

Reply:
My guess is asking numpy for 'F' order just permutes the numpy arrays sizes/strides array accordingly.